torch/_overrides.py

```diff
@@ -0,0 +1,242 @@
+"""
+Preliminary implementation of __torch_function__
+
+TODO: rewrite this in C++ for performance.
+
+NOTE: heavily inspired by NumPy's ``__array_function__`` (see:
+https://github.com/pytorch/pytorch/issues/24015 and
+https://www.numpy.org/neps/nep-0018-array-function-protocol.html
+)
+
+"""
+
+import collections
+import functools
+import textwrap
+
+# TODO: PyTorch does not have a hard dependency on NumPy, so we need
+#       to vendor this code.
+from numpy.compat._inspect import getargspec
+
+from .tensor import Tensor
+
+
+_TORCH_FUNCTION = Tensor.__torch_function__
+_TENSOR_ONLY = [Tensor]
+
+def get_overloaded_types_and_args(relevant_args):
+    """Returns a list of arguments on which to call __torch_function__.
+
+    Parameters
+    ----------
+    relevant_args : iterable of array-like
+        Iterable of array-like arguments to check for __torch_function__
+        methods.
+
+    Returns
+    -------
+    overloaded_types : collection of types
+        Types of arguments from relevant_args with __torch_function__ methods.
+    overloaded_args : list
+        Arguments from relevant_args on which to call __torch_function__
+        methods, in the order in which they should be called.
+
+    """
+    # Runtime is O(num_arguments * num_unique_types)
+    overloaded_types = []
+    overloaded_args = []
+    for arg in relevant_args:
+        arg_type = type(arg)
+        # We only collect arguments if they have a unique type, which ensures
+        # reasonable performance even with a long list of possibly overloaded
+        # arguments.
+        if (arg_type not in overloaded_types and
```

Comment:
I always find it interesting when an `O(n)` lookup is used over `O(1)`. `overloaded_types` is probably always small so this should not be a problem.

Reply:
(from reading below: order matters!)